apis/python/src/tiledbsoma/soma_array.cc

```diff
@@ -0,0 +1,506 @@
+                        reader.set_dim_points(
+                            dim, coords.cast<std::vector<std::string>>());
+                    } else if (
+                        !strcmp(arrow_schema.format, "tss:") ||
+                        !strcmp(arrow_schema.format, "tsm:") ||
+                        !strcmp(arrow_schema.format, "tsu:") ||
+                        !strcmp(arrow_schema.format, "tsn:")) {
+                        // convert the Arrow Array to int64
+                        auto pa = py::module::import("pyarrow");
+                        coords = array.attr("cast")(pa.attr("int64")()).attr("tolist")();
+                        reader.set_dim_points(
+                            dim, coords.cast<std::vector<int64_t>>());
+                    } else if (
+                        !strcmp(arrow_schema.format, "U") ||
+                        !strcmp(arrow_schema.format, "Z")) {
+                        reader.set_dim_points(
+                            dim, coords.cast<std::vector<std::string>>());
+                    } else {
+                        TPY_ERROR_LOC(
+                            "[pytiledbsoma] set_dim_points: type={} not "
+                            "supported" + 
+                            std::string(arrow_schema.format));
+                    }
+
+                    // Release arrow schema
+                    arrow_schema.release(&arrow_schema);
+                }
+            },
+            "dim"_a,
+            "py_arrow_array"_a,
+            "partition_index"_a = 0,
+            "partition_count"_a = 1)
+
+        // The following short functions are expected to be invoked when the
+        // coords are Python list/tuple, or NumPy arrays.  Arrow arrays are in
+        // the long if-else-if function above.
+        //
+        // Binding overloaded methods to templated member functions requires
+        // more effort, see:
+        // https://pybind11.readthedocs.io/en/stable/classes.html#overloaded-methods
+
+        // In an initial version of this file we had `set_dim_ranges` relying
+        // solely on type-overloading. This worked since we supported only int
+        // and string indices. In a subsequent version we are now supporting
+        // various NumPy/PyArrow types including float32, float64, int8, uint16,
+        // etc. It is an unfortunate fact that pybind11 does _not_ successfully
+        // disambiguate between float32 and float64, or between int8 and int64,
+        // etc. given that we ask it to disambiguate using not just types but
+        // std::vector of types or std::vector of std::pair of types.
+        // Experiments have shown that when both float32 and float64 are
+        // implemented with overloaded names to be differentiated solely by
+        // type, pybind11 uses the _first found_. Therefore it is necessary for
+        // us to no longer use common overloaded names.
+
+        .def(
+            "set_dim_points_string_or_bytes",
```

Comment:
Should these be `_` prefixed to indicate they are internal?

Reply:
No we've been using `SOMAArray.set_dim...` in this code for a while now so I don't want to change the syntax now.